src/util/util.js

```diff
@@ -13,60 +15,96 @@ function track(URLTemplates, variables, options) {
+    !/^[0-9]{3}$/.test(macros['ERRORCODE'])
   ) {
-    variables['ERRORCODE'] = 900;
+    macros['ERRORCODE'] = 900;
   }
 
   // Calc random/time based macros
-  variables['CACHEBUSTING'] = leftpad(
+  macros['CACHEBUSTING'] = leftpad(
     Math.round(Math.random() * 1.0e8).toString()
   );
-  variables['TIMESTAMP'] = encodeURIComponentRFC3986(new Date().toISOString());
+  macros['TIMESTAMP'] = new Date().toISOString();
 
   // RANDOM/random is not defined in VAST 3/4 as a valid macro tho it's used by some adServer (Auditude)
-  variables['RANDOM'] = variables['random'] = variables['CACHEBUSTING'];
+  macros['RANDOM'] = macros['random'] = macros['CACHEBUSTING'];
+
+  for (const macro in macros) {
+    macros[macro] = encodeURIComponentRFC3986(macros[macro]);
+  }
 
   for (const URLTemplateKey in URLArray) {
-    let resolveURL = URLArray[URLTemplateKey];
+    const resolveURL = URLArray[URLTemplateKey];
 
     if (typeof resolveURL !== 'string') {
       continue;
     }
+    const replacedUrlMacros = replaceUrlMacros(resolveURL, macros);
+    resolvedURLs.push(replacedUrlMacros);
+  }
+  return resolvedURLs;
+}
 
-    for (const key in variables) {
-      const value = variables[key];
-      const macro1 = `[${key}]`;
-      const macro2 = `%%${key}%%`;
-      resolveURL = resolveURL.replace(macro1, value);
-      resolveURL = resolveURL.replace(macro2, value);
-    }
-    resolvedURLs.push(resolveURL);
+/**
+ * Replace the macros tracking url with their value.
+ * If no value is provided for a supported macro and it exists in the url,
+ * it will be replaced by -1 as described by the VAST 4.1 iab specifications
+ *
+ * @param {String} url - Tracking url.
+ * @param {Object} macros - Object of macros to be replaced in the tracking calls
+ */
+function replaceUrlMacros(url, macros) {
+  url = replaceMacrosValues(url, macros);
+  // match any macros from the url that was not replaced
+  const remainingMacros = url.match(/[^[\]]+(?=])/g);
+  if (!remainingMacros) {
+    return url;
   }
 
-  return resolvedURLs;
+  let supportedRemainingMacros = remainingMacros.filter(
+    macro => supportedMacros.indexOf(macro) > -1
+  );
+  if (!supportedRemainingMacros) {
+    return url;
+  }
+
+  supportedRemainingMacros = supportedRemainingMacros.reduce(
+    (accumulator, macro) => {
+      accumulator[macro] = -1;
+      return accumulator;
+    },
+    {}
+  );
+  return replaceMacrosValues(url, supportedRemainingMacros);
+}
+
+/**
+ * Replace the macros tracking url with their value.
+ *
+ * @param {String} url - Tracking url.
+ * @param {Object} macros - Object of macros to be replaced in the tracking calls
+ */
+function replaceMacrosValues(url, macros) {
+  let replacedMacrosUrl = url;
+  for (const key in macros) {
+    const value = macros[key];
+    const macro1 = `[${key}]`;
+    const macro2 = `%%${key}%%`;
+    replacedMacrosUrl = replacedMacrosUrl.replace(macro1, value);
+    replacedMacrosUrl = replacedMacrosUrl.replace(macro2, value);
```

Comment:
Let use power of regex for better performance
```
// this will match [${key}] and %%${key}%%
let macroRegex = new RegExp(`(?:\[|%%)([${key}])(?:\]|%%)`, 'g');
replacedMacrosUrl = replacedMacrosUrl.replace(macroRegex, value);
```

Reply:
Not only better performance btw, with the regexp 'g' option it will replace all the occurence while a normal replace will only replace the first.

Reply:
Sure, i replaced it. It was a legacy code that i should have optimized before using  😅